modules/core/shared/src/main/scala/io/circe/PathToRoot.scala

```diff
@@ -0,0 +1,117 @@
+    prependElem(elem)
+
+  def :+(elem: PathToRoot.PathElem): PathToRoot =
+    appendElem(elem)
+}
+
+private[circe] object PathToRoot {
+  val empty: PathToRoot = PathToRoot(Vector.empty)
+
+  /**
+   * Used to describe the breadcrumbs back to the root of the JSON.
+   */
+  sealed abstract class PathElem extends Product with Serializable
+
+  object PathElem {
+    final case class ObjectKey(keyName: String) extends PathElem
+    final case class ArrayIndex(index: Int) extends PathElem
+  }
+
+  /**
+   * Convert a [[PathToRoot]] into a `String` representation of a path into a
+   * JSON object. For example, for the JSON `{"a": [1,2]}`, a cursor pointing
+   * at `2` would yield a `String`of the form `"a[1]"`.
+   */
+  def toPathString(path: PathToRoot): String = {
+    import PathElem._
+
+    if (path.value.isEmpty) {
+      ""
+    } else {
+      path.value
+        .foldLeft((new StringBuilder(path.value.size * 5), true)) {
+          case ((sb, _), ObjectKey(keyName)) =>
+            (sb.append(".").append(keyName), false)
+          case ((sb, _), ArrayIndex(index)) =>
+            (sb.append("[").append(index.toString).append("]"), false)
+        }
+        ._1
+        .toString
+    }
+  }
+
+  def fromHistory(ops: List[CursorOp]): Either[String, PathToRoot] = {
+    type F[A] = Either[String, A] // Kind Projector?
+
+    ops.reverse
+      .foldM[F, Vector[PathElem]](Vector.empty[PathElem]) {
+        case (acc :+ PathElem.ArrayIndex(n), CursorOp.MoveLeft) if n > 0 =>
+          Right(acc :+ PathElem.ArrayIndex(n - 1))
+        case (acc :+ PathElem.ArrayIndex(n), CursorOp.MoveLeft) =>
+          Left("Attempt to move beyond beginning of array in cursor history.")
+        case (acc :+ PathElem.ArrayIndex(n), CursorOp.MoveRight) if n < Int.MaxValue =>
+          Right(acc :+ PathElem.ArrayIndex(n + 1))
+        case (acc :+ PathElem.ArrayIndex(n), CursorOp.MoveRight) =>
+          Left("Attempt to move to index > Int.MaxValue in array in cursor history.")
+        case (acc :+ _, CursorOp.MoveUp) =>
+          Right(acc)
+        case (acc, CursorOp.MoveUp) =>
+          Left("Attempt to move up, but never descended in cursor history.")
+        case (acc :+ PathElem.ObjectKey(_), CursorOp.Field(name)) =>
+          Right(acc :+ PathElem.ObjectKey(name))
+        case (_, CursorOp.Field(name)) =>
+          Left("Attempt to move to sibling field, but cursor history didn't indicate we were in an object.")
+        case (acc, CursorOp.DownField(name)) =>
+          Right(acc :+ PathElem.ObjectKey(name))
+        case (acc, CursorOp.DownArray) =>
+          Right(acc :+ PathElem.ArrayIndex(0))
+        case (acc, CursorOp.DownN(n)) =>
+          Right(acc :+ PathElem.ArrayIndex(n))
+        case (acc :+ _, CursorOp.DeleteGoParent) =>
+          Right(acc)
+        case (acc :+ _, CursorOp.DeleteGoParent) =>
+          Left("Attempt to move up, but never descended in cursor history.")
```

Comment:
These are the same 2 patterns, so the 2nd one never matches?  Also i don't think the wording of the 2nd pattern matches the pattern?